File: batch_generate_reports.js
/**
 * Batch Generate Index Reports using Playwright
 * 
 * Usage:
 *   npx playwright install chromium  (first time only)
 *   node batch_generate_reports.js
 * 
 * Prerequisites:
 *   - Node.js 18+
 *   - npm install playwright
 *   - Local server running: cd docs && python3 -m http.server 8765
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const DOCS_DIR = path.join(__dirname, 'docs');
const DOWNLOADS_DIR = path.join(__dirname, 'downloads');
const REPORTS_DIR = path.join(DOCS_DIR, 'reports');
const SERVER_URL = 'http://localhost:8765';

// Ensure reports directory exists
if (!fs.existsSync(REPORTS_DIR)) fs.mkdirSync(REPORTS_DIR, { recursive: true });

// Get all CSV signal IDs
const csvFiles = fs.readdirSync(DOWNLOADS_DIR).filter(f => f.match(/^forex-forest-signals-page-(\d+)\.csv$/));
const csvIds = csvFiles.map(f => f.match(/page-(\d+)\.csv$/)[1]).sort((a, b) => parseInt(a) - parseInt(b));

// Filter out existing reports
const needed = csvIds.filter(id => !fs.existsSync(path.join(REPORTS_DIR, `index_${id}.html`)));

console.log(`Total CSVs: ${csvIds.length} | Already generated: ${csvIds.length - needed.length} | To generate: ${needed.length}`);

if (needed.length === 0) {
  console.log('✅ All reports up to date!');
  process.exit(0);
}

(async () => {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  let success = 0, fail = 0;

  for (let i = 0; i < needed.length; i++) {
    const sid = needed[i];
    const csvFile = `forex-forest-signals-page-${sid}.csv`;
    const outFile = path.join(REPORTS_DIR, `index_${sid}.html`);

    try {
      console.log(`[${i + 1}/${needed.length}] Signal #${sid} ...`);

      // Navigate to index.html
      await page.goto(SERVER_URL + '/index.html', { waitUntil: 'networkidle' });

      // Inject CSV via fetch
      const result = await page.evaluate(async (csvName) => {
        const input = document.getElementById('csvInput');
        input.value = '';
        input.files = new DataTransfer().files;

        const resp = await fetch('/downloads/' + csvName);
        if (!resp.ok) return { error: 'fetch failed', status: resp.status };
        const blob = await resp.blob();
        const file = new File([blob], csvName, { type: 'text/csv' });
        const dt = new DataTransfer();
        dt.items.add(file);
        input.files = dt.files;
        input.dispatchEvent(new Event('change', { bubbles: true }));
        return { ok: true, size: blob.size };
      }, csvFile);

      if (result.error) {
        console.log(`  ❌ CSV inject failed: ${result.error}`);
        fail++;
        continue;
      }

      // Click analyze button
      await page.click('#analyzeBtn');

      // Wait for results to appear
      await page.waitForSelector('#results', { timeout: 15000 });
      await page.waitForTimeout(3000); // Extra time for full render

      // Generate and save report
      const html = await page.evaluate(() => {
        if (typeof downloadReport === 'function') downloadReport();
        return window._reportHTML;
      });

      if (!html) {
        console.log(`  ❌ Report generation failed (no HTML)`);
        fail++;
        continue;
      }

      fs.writeFileSync(outFile, html);
      const sizeKB = (Buffer.byteLength(html) / 1024).toFixed(1);
      console.log(`  ✅ ${sizeKB} KB`);
      success++;

    } catch (err) {
      console.log(`  ❌ Error: ${err.message}`);
      fail++;
    }
  }

  await browser.close();
  console.log(`\n========================================`);
  console.log(`Batch Complete: ${success} ✅ | ${fail} ❌ | Total: ${needed.length}`);
  console.log(`========================================`);
})();

File: batch_generate_reports.py
#!/usr/bin/env python3
"""Batch generate deep analysis reports for multiple signals."""
import time, os, sys, subprocess

BASE = "/home/alvin/.openclaw/workspace/trade_strategy_analyzer"
SIGNALS = [
    "30503", "36377", "36379", "36510", "36512", "36513", "36519", "36520",
    "36655", "36656", "36657", "36658", "37850", "37851", "38641", "38663",
    "38667", "38683", "38693", "38698", "38699", "38761", "38762", "38897",
    "38900", "43024", "44452", "44453", "44459", "44465"
]

# Python script template for each signal
SCRIPT_TEMPLATE = '''
import time, os
from playwright.sync_api import sync_playwright

SIGNAL_ID = "{sig}"
BASE = "{base}"
CSV_PATH = f"{{BASE}}/samples/forex-forest-signals-page-{{SIGNAL_ID}}.csv"
OUT = f"{{BASE}}/output/index_{{SIGNAL_ID}}.html"

if not os.path.exists(CSV_PATH):
    print(f"SKIP: {{SIGNAL_ID}} - no CSV")
    exit(0)

# Check if CSV has actual trades (more than just header + deposit)
with open(CSV_PATH, "r") as f:
    lines = f.readlines()
if len(lines) < 3:
    print(f"SKIP: {{SIGNAL_ID}} - only {{len(lines)}} lines (no trades)")
    # Create a minimal placeholder report
    html_content = f"""<!DOCTYPE html>
<html><head><meta charset="UTF-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>Signal {{SIGNAL_ID}}</title></head>
<body style="max-width:900px;margin:0 auto;padding:20px;font-family:sans-serif">
<h2>Signal {{SIGNAL_ID}}</h2>
<p>No trading data available (deposit only, no trades executed).</p>
</body></html>"""
    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    with open(OUT, "w") as f:
        f.write(html_content)
    print(f"PLACEHOLDER: {{OUT}}")
    exit(0)

with open(f"{{BASE}}/index.html", "r") as f:
    html = f.read()

with sync_playwright() as p:
    br = p.chromium.launch(headless=True)
    pg = br.new_page()
    pg.route("http://localhost:8765/**", lambda route: route.fulfill(body=html, content_type="text/html"))
    pg.goto("http://localhost:8765/index.html")
    pg.wait_for_load_state("networkidle")
    pg.locator("#csvInput").set_input_files(CSV_PATH)
    time.sleep(1)
    pg.evaluate("""() => {{
        const input = document.getElementById("csvInput");
        if (input.files && input.files[0]) {{
            csvFile = input.files[0];
            document.getElementById("csvZone").classList.add("has-file");
            document.getElementById("csvZone").querySelector(".label").textContent = csvFile.name;
        }}
        document.getElementById("analyzeBtn").disabled = false;
    }}""")
    pg.locator("#analyzeBtn").click()
    for i in range(120):
        time.sleep(5)
        if pg.evaluate("""() => document.getElementById("results").innerHTML.trim() !== """""):
            break
    time.sleep(3)
    r = pg.evaluate("""() => {{
        const s = document.querySelector("style").textContent;
        const r = document.getElementById("results");
        return "<!DOCTYPE html><html><head><meta charset=\\"UTF-8\\"><meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\"><title>Trade Strategy Analysis Report</title><style>"+s+"</style></head><body style=\\"max-width:900px;margin:0 auto;padding:20px;font-family:sans-serif\\">"+r.innerHTML+"</body></html>";
    }}""")
    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    with open(OUT, "w") as f:
        f.write(r)
    br.close()
print(f"DONE: {{OUT}}")
'''

success = []
failed = []
skipped = []

for sig in SIGNALS:
    csv_path = f"{BASE}/samples/forex-forest-signals-page-{sig}.csv"
    out_path = f"{BASE}/output/index_{sig}.html"
    
    if not os.path.exists(csv_path):
        print(f"❌ {sig}: No CSV")
        failed.append(sig)
        continue
    
    if os.path.exists(out_path):
        print(f"⏭️  {sig}: Already exists")
        skipped.append(sig)
        continue
    
    script = SCRIPT_TEMPLATE.format(sig=sig, base=BASE)
    
    print(f"🔄 Processing {sig}...", flush=True)
    try:
        result = subprocess.run(
            ["python3", "-c", script],
            capture_output=True, text=True, timeout=600,
            cwd=BASE
        )
        output = result.stdout.strip() + result.stderr.strip()
        print(f"   → {output[-200:]}", flush=True)
        
        if os.path.exists(out_path):
            success.append(sig)
        else:
            failed.append(sig)
            print(f"❌ {sig}: Report not generated", flush=True)
    except subprocess.TimeoutExpired:
        failed.append(sig)
        print(f"❌ {sig}: Timeout", flush=True)
    except Exception as e:
        failed.append(sig)
        print(f"❌ {sig}: {e}", flush=True)
    
    time.sleep(2)

print(f"\n=== SUMMARY ===")
print(f"✅ Success: {len(success)} — {success}")
print(f"⏭️  Skipped: {len(skipped)} — {skipped}")
print(f"❌ Failed: {len(failed)} — {failed}")
